spi-proposal.md

```diff
@@ -0,0 +1,248 @@
+11  csActiveState         (required if csPin specified) 0: Active LOW, 1: Active HIGH
+10|12: END_SYSEX
+```
+
+#### bitOrder
+```
+LSBFIRST = 0
+MSBFIRST = 1
+```
+
+#### dataMode
+
+| mode    | clock polarity (CPOL) | clock phase (CPHA) |
+| --------|-----------------------|--------------------|
+| 0       | 0                     | 0                  |
+| 1       | 0                     | 1                  |
+| 2       | 1                     | 0                  |
+| 3       | 1                     | 1                  |
+
+
+#### maxSpeed
+
+The maximum speed of communication with the SPI device. For a SPI device rated up to
+5 MHz, use 5000000.
+
+
+### SPI_END_TRANSACTION
+
+Signals the end of a transaction. Must be sent before starting communication
+with another SPI device.
+
+```
+0:  START_SYSEX
+1:  SPI_DATA              (0x68)
+2:  SPI_END_TRANSACTION   (0x02)
+3:  channel               (HW supports multiple SPI ports. range = 0-3, default = 0)
+4:  END_SYSEX
+```
+
+### SPI_TRANSFER
+
+Write to and read from a SPI device. There are 3 ways to send and receive data
+from the SPI slave device:
+
+1. **Read/Write** For each byte written a byte is read simultaneously.
+2. **Write** Only write data (ignore any data returned by the slave device).
+3. **Read** Only read data, writing `0x00` for each byte to be read.
+
+The user can continue sending as many transfer messages as necessary, even
+alternating between read-only, write-only and read/write as necessary within
+a single SPI transaction.
+
+If a chip select (CS) pin is specified in the `SPI_BEGIN_TRANSACTION` message, it
+will be used according to the behavior specified by `pinOptions` parameter (byte 4).
+Use bits 2, 3 and 4 for reading or writing data over multiple spi transfers. Set
+bit 3 (CS_START_ONLY) for the first transfer, bit 2 (CS_DISABLE) for any middle transfers
+and finally bit 4 (CS_END_ONLY) for the last transfer. Another option, `CS_TOGGLE`
+(bit 5) allows the user to specify whether or not the CS pin should be toggled at the
+end of the message (typical SPI behavior) or between each byte sent (edge case).
+
+
+```
+0:  START_SYSEX
+1:  SPI_DATA              (0x68)
+2:  SPI_TRANSFER          (0x03)
+3:  channel               (HW supports multiple SPI ports. range = 0-3, default = 0)
+4:  pinOptions | transferMode
+                          bits 0-1: transferMode
+                            0 = read/write,
+                            1 = read-only,
+                            2 = write-only
+                            (see details in transferMode section below)
+                          bits 2-4 : pinOptions
+                            bit 2: CS_DISABLE
+                            bit 3: CS_START_ONLY
+                            bit 4: CS_END_ONLY
+                            bit 5: CS_TOGGLE
+                            (see details in pinOptions section below)
+5.  numBytes              (number of bytes to read, write or read & write)
```

Comment:
In an earlier proposal I had a `wordSize` parameter before `numBytes` and "numBytes" was actually "numWords". This is because data transfers in SPI may be of different numbers of bits. While 8 is most common (and I think is the only option the Arduino SPI implementation currently supports), it would be more scalable to make the number of bits configurable (and default to 8 bits). I think this would also open up using SPI for shift registers.